src/systems/buoyancy_engine/BuoyancyEngine.cc

```diff
@@ -77,8 +80,32 @@ class ignition::gazebo::systems::BuoyancyEnginePrivateData
 
   /// \brief mutex for protecting bladder volume and set point.
   public: std::mutex mtx;
+
+  /// \brief get fluid density
+  public: double CurrentFluidDensity(
+    EntityComponentManager &_ecm,
+    const Link &_link) const;
 };
 
+//////////////////////////////////////////////////
+double BuoyancyEnginePrivateData::CurrentFluidDensity(
+  EntityComponentManager &_ecm,
+  const Link &_link) const
+{
+  if (!this->surface.has_value())
+    return fluidDensity;
+
+  auto pose = _link.WorldPose(_ecm);
```

Comment:
An alternative to `Link::WorldPose` could be `gazebo::worldPose(this->linkEntity, _ecm)`, so you don't have to pass the link object.